### PR TITLE
Recording branch

### PR DESCRIPTION
**Pull request by Tzyinc**

### CODE REVIEW

File: MIDIchlorians/MIDIchlorians/RecorderManager.swift [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                timer.invalidate()
+                playingStarted = false
+                return
+            }
+        }
+        let playTimeInt = toPlay[0].0
+        if playClock >= playTimeInt {
+            for value in toPlay {
+                //TODO: Call pad to be played here
+                print("page", value.1.0, "index", value.1.1)
```
Comment (Tzyinc) [inline MIDIchlorians/MIDIchlorians/RecorderManager.swift]: need @ngzhian 's help to play pad from page

Comment (ngzhian) [reply MIDIchlorians/MIDIchlorians/RecorderManager.swift]: can make a `RecordPlaybackDelegate` protocol with a play method, then I'll make `GridController` implement it

File: MIDIchlorians/MIDIchlorians/RecorderManager.swift [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+class RecorderManager {
+    static let instance = RecorderManager()
+
+    //for recording
+    private var recordingStarted = false
+    private var timeTracker: TimeTracker
+
+    //for playback
+    private var playBack: PlayBackRetriever
+    private let TIMEINTERVAL: TimeInterval = 1/60
```
Comment (ngzhian) [inline MIDIchlorians/MIDIchlorians/RecorderManager.swift]: Throw this into Config?

File: MIDIchlorians/MIDIchlorians/TimeTracker.swift [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+//  Copyright © 2017 nus.cs3217.a0118897. All rights reserved.
+//
+
+import Foundation
+
+struct TimeTracker {
+    private var timeIndexArr: [(TimeInterval, (Int, IndexPath))]
+    private let start: Date
+    //the 'fps' at which we will play the recording
+    private let PLAYBACKACCURACY: TimeInterval = 1/60
```
Comment (ngzhian) [inline MIDIchlorians/MIDIchlorians/TimeTracker.swift]: throw into Config

File: MIDIchlorians/MIDIchlorians/PlayBackRetriever.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+struct PlayBackRetriever {
+    private var timeIndexArr: [(TimeInterval, (Int, IndexPath))]
+    init(timeIndexArr: [(TimeInterval, (Int, IndexPath))]) {
+        self.timeIndexArr = timeIndexArr
+    }
+
+    //returns empty arr if there's nothing left
+    //otherwise, returns one or more tuples
+    mutating func getNextPads() -> [(TimeInterval, (Int, IndexPath))] {
```
Comment (ngzhian) [inline MIDIchlorians/MIDIchlorians/PlayBackRetriever.swift:19]: what is this method trying to do actually, is it returning all pads that have the same time interval?

File: MIDIchlorians/MIDIchlorians/RecorderManager.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    func startRecord() {
+        timeTracker = TimeTracker()
+        recordingStarted = true
+    }
+
+    func stopRecord() {
+        recordingStarted = false
+    }
+
+    func recordPad (forPage pageNum: Int, forIndex indexPath: IndexPath) {
```
Comment (ngzhian) [inline MIDIchlorians/MIDIchlorians/RecorderManager.swift]: nit: no space between method name and parentheses

**Review by ngzhian — CHANGES_REQUESTED** [review]

File: MIDIchlorians/MIDIchlorians/RecorderManager.swift (hunk 5 above)
Comment (Tzyinc) [reply MIDIchlorians/MIDIchlorians/RecorderManager.swift]: good catch

File: MIDIchlorians/MIDIchlorians/PlayBackRetriever.swift (hunk 4 above)
Comment (Tzyinc) [reply MIDIchlorians/MIDIchlorians/PlayBackRetriever.swift:19]: yes, it's returning all pads with the same time interval, to be called when the current pads have been played (think of it like buffering the next pads to play)

**Review by ngzhian — APPROVED** [review]

**Comment by ngzhian** [comment]
Fix conflicts and can merge :) Nice!

**Pull request merged**